lightning/src/routing/router.rs

```diff
@@ -242,16 +248,96 @@ impl_writeable_tlv_based!(RouteHop, {
+	///
+	/// [`BlindedPath`]: crate::blinded_path::BlindedPath
+	pub intro_node_scid: u64,
+}
+
+impl_writeable_tlv_based!(BlindedTail, {
+	(0, path, required),
+	(2, fee_msat, required),
+	(4, cltv_expiry_delta, required),
+	(6, final_value_msat, required),
+	(8, intro_node_scid, required),
+});
+
+/// A path in a [`Route`] to the payment recipient.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Path {
+	/// The list of unblinded hops in this [`Path`]. May be empty if a blinded path is present,
+	/// otherwise the last hop is the destination and this must be at least length one.
+	pub hops: Vec<RouteHop>,
+	/// The blinded path at which this path terminates, if we're sending to one, and its metadata.
+	pub blinded_tail: Option<BlindedTail>,
+}
+
+impl Path {
+	/// Returns the total amount of fees paid on this [`Path`].
+	pub fn fee_msat(&self) -> u64 {
+		match &self.blinded_tail {
+			Some(tail) => self.hops.iter().map(|hop| hop.fee_msat).sum::<u64>() + tail.fee_msat,
+			None =>
+				self.hops.split_last().map_or(0,
+					|(_, path_prefix)| path_prefix.iter().map(|hop| hop.fee_msat).sum())
+			}
+	}
+
+	/// Returns the total amount paid on this [`Path`], excluding the fees.
+	pub fn final_value_msat(&self) -> u64 {
+		match &self.blinded_tail {
+			Some(blinded_tail) => blinded_tail.final_value_msat,
+			None => self.hops.last().map_or(0, |hop| hop.fee_msat)
+		}
+	}
+
+	/// The length of this path, including the hops of its [`Path::blinded_tail`], if any.
+	pub fn len(&self) -> usize {
+		self.hops.len() + self.blinded_tail.as_ref().map_or(0, |tail| tail.path.blinded_hops.len())
+	}
+
+	pub(crate) fn first_hop_node_id(&self) -> PublicKey {
+		self.hops.first().map_or_else(
+			|| self.blinded_tail.as_ref().expect("Path had 0 hops").path.introduction_node_id.clone(),
+			|hop| hop.pubkey)
+	}
+
+	pub(crate) fn first_hop_scid(&self) -> u64 {
+		self.hops.first().map_or_else(
+			|| self.blinded_tail.as_ref().expect("Path had 0 hops").intro_node_scid,
+			|hop| hop.short_channel_id)
+	}
+}
+
 /// A route directs a payment from the sender (us) to the recipient. If the recipient supports MPP,
 /// it can take multiple paths. Each path is composed of one or more hops through the network.
 #[derive(Clone, Hash, PartialEq, Eq)]
 pub struct Route {
-	/// The list of paths taken for a single (potentially-)multi-part payment. The pubkey of the
-	/// last [`RouteHop`] in each path must be the same. Each entry represents a list of hops, NOT
-	/// INCLUDING our own, where the last hop is the destination. Thus, this must always be at
-	/// least length one. While the maximum length of any given path is variable, keeping the length
-	/// of any path less or equal to 19 should currently ensure it is viable.
-	pub paths: Vec<Vec<RouteHop>>,
+	/// The list of [`Path`]s taken for a single (potentially-)multi-part payment. Each entry
+	/// represents a list of hops, NOT INCLUDING our own, where the last hop is either the recipient
```

Comment:
> NOT INCLUDING our own

Hmm... a hop to me is a channel so this wording doesn't seem accurate.

Reply:
I reworked these docs in a fixup